runtime/jitserver_launcher/jitserver.c

```diff
@@ -0,0 +1,400 @@
+#endif
+
+	/* Consume any trailing '/' */
+	if (path[pathLen - 1] == DIR_SEPARATOR_CHAR) {
+		path[pathLen - 1] = '\0';
+	}
+
+	/* Now consume the last directory if present */	
+	lastLevel = strrchr(path, DIR_SEPARATOR_CHAR);
+	if (NULL != lastLevel) {
+		*lastLevel = '\0';	
+	}
+
+#if defined(JITSERVER_LAUNCHER_DEBUG)
+	fprintf(stdout, "truncateOneLevel> on exit, path: %s\n", path);
+#endif
+}
+
+static int32_t 
+append(char *buffer, int32_t bufLen, char *str)
+{
+	if (strlen(buffer) + strlen(str) + 1 > bufLen) {
+		return -1;
+	} else {
+		strcat(buffer, str);
+		return 0;
+	}
+}
+
+static int32_t
+isDir(const char *path) {
+	int32_t rc = 0;
+	struct stat statBuf = {0};
+
+	rc = stat(path, &statBuf);
+	if ((0 == rc) && (S_ISDIR(statBuf.st_mode))) {
+		return 1;
+	}
+	return 0;
+}
+
+static int32_t
+isRegularFile(const char *path) {
+	int32_t rc = 0;
+	struct stat statBuf = {0};
+
+	rc = stat(path, &statBuf);
+	if ((0 == rc) && (S_ISREG(statBuf.st_mode))) {
+		return 1;
+	}
+	return 0;
+}
+
+static int32_t
+getJvmPath(char *pathBuffer, int32_t pathBufLen) {
+	int32_t rc = 0;
+
+ 	/* pathBuffer can be either <java_home> or <java_home>/jre */
+	rc = checkLastLevel(pathBuffer, "jre");
+	if (1 == rc) {
+		/* we are already at <java_home>/jre */
+		rc = append(pathBuffer, pathBufLen, "/lib/" OPENJ9_ARCH_DIR);
+	} else {
+#if JAVA_SPEC_VERSION >= 11 
+		rc = append(pathBuffer, pathBufLen, "/lib/" OPENJ9_ARCH_DIR);
+#else
+		rc = append(pathBuffer, pathBufLen, "/jre/lib/" OPENJ9_ARCH_DIR);
+
+#endif /* JAVA_SPEC_VERSION */
+	}
+	if (0 != rc) {
+		fprintf(stderr, "Path exceeded maximum allowed length of %d\n", OPENJ9_MAX_PATH);
+		goto _end;
+	}
+
+	/* try "compressedrefs" first */
+	rc = append(pathBuffer, pathBufLen, "/" OPENJ9_CR_JVM_DIR);
+        if (0 != rc) {
+                fprintf(stderr, "Path exceeded maximum allowed length of %d\n", OPENJ9_MAX_PATH);
+                goto _end;
+        }
+	rc = isDir(pathBuffer);
+	if (0 == rc) {
+		/* "compressedrefs" is not present, try "noncr" now */
+		truncateOneLevel(pathBuffer);
+		rc = append(pathBuffer, pathBufLen, "/" OPENJ9_NOCR_JVM_DIR);
+		if (0 != rc) {
+			fprintf(stderr, "Path exceeded maximum allowed length of %d\n", OPENJ9_MAX_PATH);
+			goto _end;
+		}
+		rc = isDir(pathBuffer);
+		if (0 == rc) {
+			fprintf(stderr, "Failed to find any recognized JVM\n");
+			rc = -1;
+			goto _end;
+		}
+	}
+#if defined(JITSERVER_LAUNCHER_DEBUG)
+	fprintf(stdout, "getJvmPath> path: %s\n", pathBuffer);
+#endif
+	rc = 0; /* success */
+
+_end:
+	return rc;
+}
+
+static char * 
+getJvmLibPath(void)
+{
+	char *jvmLibPath = NULL;
+	char *selfPath = NULL;
+	int32_t rc = 0;
+
+	selfPath = getSelfPath();
+	if (NULL == selfPath) {
+		goto _end;
+	}
+	jvmLibPath = malloc(OPENJ9_MAX_PATH);
+	if (NULL == jvmLibPath) {
+		fprintf(stderr, "Failed to allocate memory for jvm library\n");
+		free(selfPath);
+		goto _end;
+	}
+	strcpy(jvmLibPath, selfPath);
+	free(selfPath);
```

Comment:
Only one of `jvmLibPath` or `selfPath` is required.

Also, this computation and the collection of support functions is needlessly complicated.